From dafd7b539d5914f708dc2e4b35041a68f743a9dc Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Mon, 24 Jul 2023 15:18:08 +0800
Subject: [PATCH 19/53] [Huawei] mm/sharepool: Delete the state machine in
 sp_k2u_context

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Use the return value of sp_remap_kva_to_vma() to indicate the coredump
state, just like that in sp_alloc.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 69 +++++++++++++++++++++++--------------------------
 1 file changed, 32 insertions(+), 37 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 9647ecf1ccd3..4ea51ec5dd98 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -860,21 +860,17 @@ struct sp_alloc_context {
 static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 				    unsigned long prot, struct sp_alloc_context *ac);
 
-#define K2U_NORMAL	0
-#define K2U_COREDUMP	1
-
 struct sp_k2u_context {
 	unsigned long kva;
 	unsigned long kva_aligned;
 	unsigned long size;
 	unsigned long size_aligned;
 	unsigned long sp_flags;
-	int state;
 	enum spa_type type;
 };
 
-static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
-					unsigned long prot, struct sp_k2u_context *kc);
+static int sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
+			       unsigned long prot, struct sp_k2u_context *kc);
 
 static void free_sp_group_id(int spg_id)
 {
@@ -1468,8 +1464,6 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	spin_lock(&sp_area_lock);
 
 	list_for_each_entry(spa, &spg->spa_list, link) {
-		unsigned long addr;
-
 		__sp_area_drop_locked(prev);
 		prev = spa;
 
@@ -1481,9 +1475,11 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		spin_unlock(&sp_area_lock);
 
 		if (spa->type == SPA_TYPE_K2SPG && spa->kva) {
-			addr = sp_remap_kva_to_vma(spa, mm, prot, NULL);
-			if (IS_ERR_VALUE(addr))
-				pr_warn("add group remap k2u failed %ld\n", addr);
+			ret = sp_remap_kva_to_vma(spa, mm, prot, NULL);
+			if (ret) {
+				pr_warn("add group remap k2u failed %d\n", ret);
+				ret = 0;
+			}
 
 			spin_lock(&sp_area_lock);
 			continue;
@@ -2415,29 +2411,28 @@ static unsigned long __sp_remap_get_pfn(unsigned long kva)
 }
 
 /* when called by k2u to group, always make sure rw_lock of spg is down */
-static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
-					unsigned long prot, struct sp_k2u_context *kc)
+static int sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
+			       unsigned long prot, struct sp_k2u_context *kc)
 {
-	struct vm_area_struct *vma;
+	int ret = 0;
 	unsigned long ret_addr;
+	struct vm_area_struct *vma;
 	unsigned long populate = 0;
-	int ret = 0;
 	unsigned long addr, buf, offset;
 	unsigned long kva = spa->kva;
 
 	down_write(&mm->mmap_lock);
 	if (unlikely(mm->core_state)) {
+		up_write(&mm->mmap_lock);
 		pr_err("k2u mmap: encountered coredump, abort\n");
-		ret_addr = -EBUSY;
-		if (kc)
-			kc->state = K2U_COREDUMP;
-		goto put_mm;
+		return SP_SKIP_ERR;
 	}
 
 	ret_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
 	if (IS_ERR_VALUE(ret_addr)) {
-		pr_debug("k2u mmap failed %lx\n", ret_addr);
-		goto put_mm;
+		up_write(&mm->mmap_lock);
+		pr_info("k2u mmap failed %lx\n", ret_addr);
+		return (int)ret_addr;
 	}
 
 	/* This should not fail because we hold the mmap_lock sicne mmap */
@@ -2446,10 +2441,10 @@ static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *
 		ret = remap_vmalloc_hugepage_range(vma, (void *)kva, 0);
 		if (ret) {
 			do_munmap(mm, ret_addr, spa_size(spa), NULL);
+			up_write(&mm->mmap_lock);
 			pr_debug("remap vmalloc hugepage failed, ret %d, kva is %lx\n",
 				 ret, (unsigned long)kva);
-			ret_addr = ret;
-			goto put_mm;
+			return ret;
 		}
 		vma->vm_flags |= VM_IO | VM_PFNMAP | VM_DONTEXPAND | VM_DONTDUMP;
 	} else {
@@ -2461,9 +2456,9 @@ static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *
 					__pgprot(vma->vm_page_prot.pgprot));
 			if (ret) {
 				do_munmap(mm, ret_addr, spa_size(spa), NULL);
+				up_write(&mm->mmap_lock);
 				pr_err("remap_pfn_range failed %d\n", ret);
-				ret_addr = ret;
-				goto put_mm;
+				return ret;
 			}
 			offset += PAGE_SIZE;
 			buf += PAGE_SIZE;
@@ -2471,10 +2466,9 @@ static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *
 		} while (offset < spa_size(spa));
 	}
 
-put_mm:
 	up_write(&mm->mmap_lock);
 
-	return ret_addr;
+	return 0;
 }
 
 static bool vmalloc_area_set_flag(unsigned long kva, unsigned long flags)
@@ -2499,10 +2493,11 @@ static bool vmalloc_area_set_flag(unsigned long kva, unsigned long flags)
  */
 static void *sp_make_share_kva_to_spg(struct sp_k2u_context *kc, struct sp_group *spg)
 {
+	int mmap_ret;
+	int ret = -EINVAL;
 	struct sp_area *spa;
 	struct mm_struct *mm;
 	struct sp_group_node *spg_node;
-	unsigned long ret_addr = -ENODEV;
 
 	down_read(&spg->rw_lock);
 	spa = sp_alloc_area(kc->size_aligned, kc->sp_flags, spg, kc->type, current->tgid, 0);
@@ -2516,27 +2511,27 @@ static void *sp_make_share_kva_to_spg(struct sp_k2u_context *kc, struct sp_group
 	spa->kva = kc->kva_aligned;
 	list_for_each_entry(spg_node, &spg->proc_head, proc_node) {
 		mm = spg_node->master->mm;
-		kc->state = K2U_NORMAL;
-		ret_addr = sp_remap_kva_to_vma(spa, mm, spg_node->prot, kc);
-		if (IS_ERR_VALUE(ret_addr)) {
-			if (kc->state == K2U_COREDUMP)
+		mmap_ret = sp_remap_kva_to_vma(spa, mm, spg_node->prot, kc);
+		if (mmap_ret) {
+			if (mmap_ret == SP_SKIP_ERR)
 				continue;
-			pr_err("remap k2u to spg failed %ld\n", ret_addr);
+			pr_err("remap k2u to spg failed %d\n", mmap_ret);
 			__sp_free(spg, spa->va_start, spa_size(spa), mm);
-			goto out;
+			ret = mmap_ret;
+			break;
 		}
+		ret = mmap_ret;
 	}
 
-out:
 	up_read(&spg->rw_lock);
-	if (!IS_ERR_VALUE(ret_addr)) {
+	if (!ret) {
 		sp_update_process_stat(current, true, spa);
 		vmalloc_area_set_flag(kc->kva_aligned, VM_SHAREPOOL);
 	}
 
 	__sp_area_drop(spa);
 
-	return (void *)ret_addr;
+	return ret ? ERR_PTR(ret) : (void *)spa->va_start;
 }
 
 static int sp_k2u_prepare(unsigned long kva, unsigned long size,
-- 
2.17.1

